Add multi-node-multi-gpu Logistic Regression in C++

lijinf2 opened the pull request:
This PR enables multi-node-multi-gpu Logistic Regression and it mostly reuses existing codes (i.e. GLMWithData and min_lbfgs) of single-GPU Logistic Regression. No change to any existing codes. 

Added Pytest code for Spark cluster and the tests run successfully with 2 GPUs on a random dataset. The coef_ and intercept_ are the same as single-GPU cuml.LogisticRegression.fit. Pytest code can be found here: https://github.com/lijinf2/spark-rapids-ml/blob/lr/python/tests/test_logistic_regression.py

wbo4958 commented on cpp/include/cuml/linear_model/qn_mg.hpp:
```diff
@@ -0,0 +1,35 @@
+namespace GLM {
+namespace opg {
+
+void toy(const raft::handle_t &handle, 
```
seem this API is not used?

lijinf2 replied:
deleted

wbo4958 commented on cpp/include/cuml/linear_model/qn_mg.hpp:
```diff
@@ -0,0 +1,35 @@
+#include <raft/core/comms.hpp>
```
copyright?

wbo4958 commented on cpp/include/cuml/linear_model/qn_mg.hpp:
```diff
@@ -0,0 +1,35 @@
+};  // namespace GLM
+};  // namespace ML
+
+
```
please delete extra empty lines.

wbo4958 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,137 @@
+#include <raft/core/comms.hpp>
```
copyright

wbo4958 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,137 @@
+                      SimpleDenseMat<T>& G,
+                      const SimpleMat<T>& X,
+                      const SimpleDenseMat<T>& dZ,
+                      bool setZero,
```
I guess we need to keep the same naming convention?

lijinf2 commented:
Rebased with the latest branch-23.08.

cjnolet commented on cpp/include/cuml/linear_model/qn_mg.hpp:
```diff
@@ -0,0 +1,43 @@
+namespace GLM {
+namespace opg {
+
+void qnFit(const raft::handle_t &handle, 
```
Doxygen would be helpful here for public APIs.

lijinf2 replied:
Added.

cjnolet commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+}
+
+template <typename T, class GLMObjective>
+struct GLMWithDataMG : ML::GLM::detail::GLMWithData<T, GLMObjective> {
```
Could we have a doxygen here just to help future developers?

lijinf2 replied:
added, please check!

cjnolet commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,126 @@
+
+}
+
+void qnFit(const raft::handle_t &handle, 
```
Doxygen could be helpful here. As long as it's here or int he header, one place is fine.

lijinf2 replied:
added to qn_mg.hpp

cjnolet commented on python/cuml/linear_model/logistic_regression_mg.pyx:
```diff
@@ -0,0 +1,198 @@
+#
+# Copyright (c) 2019-2022, NVIDIA CORPORATION.
```
Please only use the current year for new files.

cjnolet commented on cpp/include/cuml/linear_model/qn_mg.hpp:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.
```
Please only use the current year for new files

cjnolet commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.
```
Please only use the current year for new files.

cjnolet commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.
```
Please only use the current year for new files.

cjnolet commented on python/cuml/linear_model/logistic_regression_mg.pyx:
```diff
@@ -0,0 +1,198 @@
+# the cdef was copied from cuml.linear_model.qn
+cdef extern from "cuml/linear_model/glm.hpp" namespace "ML::GLM" nogil:
+
+    cdef enum qn_loss_type "ML::GLM::qn_loss_type":
```
Since this isn't MG specific, we should centralize it so the SG and MG can use the same one.

lijinf2 replied:
This will affect single-GPU logistic_regression.pyx. Is it ok to do it in the next PR to avoid this one being too long?

dantegd replied:
I agree with the logic of adding this change in a follow up, but could you open a GH issue capturing this (and any other small) follow up items?

cjnolet replied:
Also, can you add a little todo here and post the number of the github issue? This helps us map it back to the code so it doesn't get lost/forgotten.

cjnolet commented on python/cuml/linear_model/logistic_regression_mg.pyx:
```diff
@@ -0,0 +1,198 @@
+        QN_LOSS_ABS      "ML::GLM::QN_LOSS_ABS"
+        QN_LOSS_UNKNOWN  "ML::GLM::QN_LOSS_UNKNOWN"
+
+    cdef struct qn_params:
```
Same here- this isn't mg specific so we should use the SG version

cjnolet commented on python/cuml/linear_model/logistic_regression_mg.pyx:
```diff
@@ -0,0 +1,198 @@
+        int n_ranks) except +
+
+
+class LogisticRegressionMG(LogisticRegression):
```
My suggestion is to use the design from `LinearRegressionMG` as much as possible to model the relationship between `LogisticRegression` and `LogisticRegressionMG`. There are some more classes and mixins that can reduce boilerplate further.

lijinf2 replied:
Find it a bit challenging to inherit mixins, for two reasons:

(1) min_lbfgs accepts float*.  mixins _fit accepts vector<float*>
(2) min_lbfgs supports multiple classes so coef_ can be a matrix.  mixins assumes coef_ to be a vector of D + pams.fit_intercept length. 

Any ideas to simplify the implementation?

lijinf2 replied:
(3) Another issue pops up that relates to mixins._fit and (1). It seems mixins.fit converts input vectors and labels into a special type of points (i.e. X_arg, y_arg) using opg.build_data_t(X_arys). Is there s way to convert X_arg and y_arg to float* (min_lbfgs requires the float*)?

lijinf2 replied:
Combined the design from 'LinearRegressionMG' and singe-gpu 'LogisticRegression'. Specifically:
(1) Got self._coef initialization support LinearRegressionMG style (i.e. 1-D array) and LogisticRegression style (i.e. 2-D array). 
(2) Converted type of input vectors and input labels) from LienarRegressionMG style (vector of Matrix:Data<float> type) to LogisticRegression style (float*). 
(3) set n_classes to 2 in this PR and will work on multiple class in following PRs.

cjnolet commented on python/cuml/linear_model/logistic_regression_mg.pyx:
```diff
@@ -0,0 +1,198 @@
+            self.solver_model._coef_ = CumlArray.zeros(
+                coef_size, dtype=self.dtype, order='C')
+
+    def fit(self, X, y, rank, n_ranks, n_samples, n_classes, convert_dtype=False) -> "LogisticRegressionMG":
```
As mentioned above, I think this could be simplified further to look more like [this](https://github.com/rapidsai/cuml/blob/00bdad3cae107ce0021739dc11860e9629af738f/python/cuml/linear_model/linear_regression_mg.pyx#L74).

lijinf2 replied:
Revised the code and the _fit now looks exactly the same. 
fit looks the same except for one additional argument (i.e. order='F'). This is because `LinearRegressionMG` uses 'F' order but `cuml.LogisticRegression` uses 'C' order.

cjnolet reviewed (changes requested):
Thanks for this new feature @lijinf2! This is going to be useful not just in Spark but in Dask and other places as well! 

I think overall the C++ code is structured pretty well. There are some general details around the communicator and handle, documentation, and input validation that should be done. 

I think the cython can be simplified further and the Dask piece should be pretty straightforward to add. The [Dask Linear Regression Wrapper](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/dask/linear_model/linear_regression.py) should be helpful there. Also, as discussed off line, we will need tests and the [Dask Linear Regression Pytests](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/tests/dask/test_dask_linear_regression.py) should be helpful there. More specifically, you should be able to use [Dask make_classification](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/dask/datasets/classification.py) data generator for that. 

Thanks again for this contribution!

lijinf2 commented:
> Thanks for this new feature @lijinf2! This is going to be useful not just in Spark but in Dask and other places as well!
> 
> I think overall the C++ code is structured pretty well. There are some general details around the communicator and handle, documentation, and input validation that should be done.
> 
> I think the cython can be simplified further and the Dask piece should be pretty straightforward to add. The [Dask Linear Regression Wrapper](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/dask/linear_model/linear_regression.py) should be helpful there. Also, as discussed off line, we will need tests and the [Dask Linear Regression Pytests](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/tests/dask/test_dask_linear_regression.py) should be helpful there. More specifically, you should be able to use [Dask make_classification](https://github.com/rapidsai/cuml/blob/branch-23.08/python/cuml/dask/datasets/classification.py) data generator for that.
> 
> Thanks again for this contribution!

Thanks so much for reviewing the PR! @cjnolet Your comments and links are very helpful and I will start working on the revision following your suggestions. One thing I feel uncertain relates to num_classes calculation. Is there an existing C++ API to calculate num_classes from labels across GPUs? This will help move num_classes variable to C++ implementation and reduce one argument in Cython fit wrapper.

cjnolet commented:
> Is there an existing C++ API to calculate num_classes from labels across GPUs?

I don't know if we have a way to do this on mutliple GPUs yet. We could use an `nunique` primitive and potentially do a distributed reduction. Maybe something we should think about a little more. Since time is important here, if we can't think of anything very soon, I would suggest we create a github issue for it and reference the github issue in a TODO in the code so we don't lose sight of it. I'll be the first to say that we are in need of a nice collection of mnmg primitives that are fully c++.

cjnolet commented:
/ok to test

lijinf2 commented:
> > Is there an existing C++ API to calculate num_classes from labels across GPUs?
> 
> I don't know if we have a way to do this on mutliple GPUs yet. We could use an `nunique` primitive and potentially do a distributed reduction. Maybe something we should think about a little more. Since time is important here, if we can't think of anything very soon, I would suggest we create a github issue for it and reference the github issue in a TODO in the code so we don't lose sight of it. I'll be the first to say that we are in need of a nice collection of mnmg primitives that are fully c++.

Figured out it is possible with C++ "raft::label::detail::getUniquelabel" and comm.allgather". But it turns out the python wrapper logistic_regression_mg.pyx also needs this num_classes on the python side to initialize coef_. So a wrapper of the C++ implementation is required. 

Maybe in this PR we can assume num_classes to be 2 and remove the num_class from Cython fit argument list. In a future PR we can implement the C++ num_class calculation and a wrapper to use it in logistic_regression_mg.pyx .

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+        raft::update_host(&tmp_host, dev_scalar, 1, stream);
+        tmp_host = tmp_host * (*this->y).len / this->n_samples;
+        lossVal.fill(tmp_host, stream);
+        raft::interruptible::synchronize(stream);
```
Deleted the sync statement. Thanks!

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+        communicator.allreduce(dev_scalar, dev_scalar, 1,                          
+            raft::comms::op_t::SUM, stream);                                                                 
+        raft::resource::sync_stream(*(this->handle_p));    
+        raft::interruptible::synchronize(stream);
```
Removed the sync_stream() and synchronize()
Added communicator.sync().

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+
+        communicator.allreduce(G.data, G.data, this->C * this->dims,                          
+            raft::comms::op_t::SUM, stream);                                                                 
+        raft::resource::sync_stream(*(this->handle_p));    
```
Replaced with comms.sync().

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+        raft::interruptible::synchronize(stream);
+
+        linearBwdMG(lossFunc->handle, G, *(this->X), *(this->Z), false, n_samples, n_ranks);    // linear part: backward pass
+        raft::interruptible::synchronize(stream);
```
Removed the sync.

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+        float loss_host;
+        raft::update_host(&loss_host, dev_scalar, 1, stream);
+
+        raft::interruptible::synchronize(stream);
```
done.

lijinf2 commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,126 @@
+  // prepare regularizer regularizer_obj
+  ML::GLM::detail::LogisticLoss<T> loss_func(handle, D, pams.fit_intercept);
+  T l1 = pams.penalty_l1;
+  T l2 = pams.penalty_l2;
```
Added a switch statement to restrict the pams.loss to be "logistic".

cjnolet replied:
We should probably also do an assertion or at least print a warning if there's an l1 set by the user at all, to avoid confusion if they have that expectation for any reason.

lijinf2 commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,153 @@
+        G.fill(0, stream);
+        reg->reg_grad(dev_scalar, G, W, lossFunc->fit_intercept, stream);
+        float reg_host;
+        raft::update_host(&reg_host, dev_scalar, 1, stream);
```
Yeah. Added a comment to explain no syncing.

lijinf2 commented:
@cjnolet Thanks so much for the comments! They are very helpful. After addressing the comments, I find the implementation of dask class (cuml.dask.linear_model.LogisticRegression) and pytests become much simpler. Would like to have your review again when you have time. 

The revised PR mainly includes two new files (dask/linear_model/LogisticRegression) and (tests/dask/test_dask_logistic_regression.py). The added pytests get passed in my workstation of 2-GPUs. Will need to run the full pytests of cuml.

dantegd commented:
/ok to test

lijinf2 commented:
/ok to test

csadorf commented:
/ok to test

dantegd commented:
/ok to test

dantegd commented on python/cuml/dask/common/base.py:
```diff
@@ -407,12 +407,13 @@ def _fit(self, model_func, data):
             ]
         )
 
+        fit_func = self._func_fit
```
This seems mostly a stylistic change, or was there a particular reason that prompted this change?

lijinf2 replied:
yeah, the change allows overriding of the _func_fit function.  This is required by class dask.linear_model.LogisticRegression(SyncFitMixinLinearModel), which overrides the function in order to reuse SyncFitMixinLinearModel.fit().

dantegd commented on python/cuml/tests/dask/test_dask_logistic_regression.py:
```diff
@@ -141,3 +141,73 @@ def imp():
+        LogisticRegression as cumlLBFGS_dask,
+    )
+
+    n_info = 5
```
Small question, why choice of 5? If there's a reason, would recommend adding a comment about it.

lijinf2 replied:
The variable seemed required by dataset generator. Added the reference.

dantegd commented on python/cuml/tests/dask/test_dask_logistic_regression.py:
```diff
@@ -141,3 +141,73 @@ def imp():
+
+    client.run(imp)
+
+    import numpy as np
```
Numpy has already been imported in this file

dantegd commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,188 @@
+    handle, input_data, input_desc, labels, coef, pams, X_col_major, n_classes, f, num_iters);
+}
+
+/*
+void qnFit(const raft::handle_t &handle,
+           const qn_params& pams,
+           float* X,
+           bool X_col_major,
+           float *y,
+           int N,
+           int D,
+           int C,
+           float* w0,
+           float* f,
+           int* num_iters,
+           int n_samples,
+           int rank,
+           int n_ranks)
+{
+  qnFit_impl<float, int>(
+    handle,
+    pams,
+    X,
+    X_col_major,
+    y,
+    N,
+    D,
+    C,
+    w0,
+    f,
+    num_iters,
+    n_samples,
+    rank,
+    n_ranks);
+}
+*/
```
Need to remove commented code?

dantegd reviewed (changes requested):
Just had some small comments, overall things look in very good shape

lijinf2 commented:
@dantegd Thanks so much for helping improve the code. I have submitted the revised code and had rebased the latest. Also, created two followup issues: [Support multiple classes](url[)](https://github.com/rapidsai/cuml/issues/5501) and [Reuse SG version qn_params and qn_loss_type](https://github.com/rapidsai/cuml/issues/5502).

dantegd commented:
/ok to test

dantegd reviewed (approved):
changes look good from my side

cjnolet commented on cpp/src/glm/qn/glm_base_mg.cuh:
```diff
@@ -0,0 +1,166 @@
+template <typename T, class GLMObjective>
+struct GLMWithDataMG : ML::GLM::detail::GLMWithData<T, GLMObjective> {
+  /**
+   * @brief Aggregates local gradient vectors and loss values from local training data. This
```
Can you put the doxygen on the actual class please?

cjnolet commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,154 @@
+{
+  switch (pams.loss) {
+    case QN_LOSS_LOGISTIC: {
+      ASSERT(C == 2, "qn_mg.cu: logistic loss invalid C");
```
Since this solver would normally support more loss functions, can you add a quick note for the user here that only the LOGISTIC loss is supported currently? Also, please use `RAFT_EXPECTS` instead of `ASSERT`  so that we have consistent messaging.

cjnolet commented on cpp/src/glm/qn_mg.cu:
```diff
@@ -0,0 +1,154 @@
+                T* f,
+                int* num_iters)
+{
+  ASSERT(input_data.size() == 1, "qn_mg.cu currently does not accept more than one input matrix");
```
Here too, please use `RAFT_EXPECTS` everywhere instead of `ASSERT`

cjnolet reviewed (changes requested):
Implementation looks great and it's coming together. My remaining comments are very minor at this point. I think we'll be able to get this into 23.08.

lijinf2 commented:
> Implementation looks great and it's coming together. My remaining comments are very minor at this point. I think we'll be able to get this into 23.08.

Thanks so much for reviewing again! I have revised the PR and pushed the latest.

cjnolet commented:
/ok to test

cjnolet reviewed (approved):
LGTM. Thanks Tamas!

cjnolet commented:
/merge

cjnolet reviewed (changes requested):
Sorry @lijinf2, that approval was premature and meant for a different PR. I think this will be getting approved shortly, though.

lijinf2 commented:
/ok to test

lijinf2 commented:
The checks seem fail at irrelevant test cases(e.g. error "FAILED test_hdbscan.py::test_all_points_membership_vectors_circles[1000-knn-leaf-0-True-0.5-500-5-1000] - TypeError: 'numpy.float64' object cannot be interpreted as an integer"). Any idea? I can get test_hdbscan.py passed locally on my workstation. Perhaps rerun "/ok to test" to repeat the error.

csadorf commented:
> The checks seem fail at irrelevant test cases(e.g. error "FAILED test_hdbscan.py::test_all_points_membership_vectors_circles[1000-knn-leaf-0-True-0.5-500-5-1000] - TypeError: 'numpy.float64' object cannot be interpreted as an integer"). Any idea? I can get test_hdbscan.py passed locally on my workstation. Perhaps rerun "/ok to test" to repeat the error.

We are currently experiencing some CI issues due to changes in our dependencies (see https://github.com/rapidsai/cuml/issues/5514). Once those are resolved, we can rerun tests here.

cjnolet commented:
/merge
